### PR TITLE
Note how Solidity types map to ABI types

### LINKED ISSUE #4950: Move mapping from solidity types to ABI types in the documentation.
Came up in https://github.com/ethereum/solidity/pull/4911#discussion_r216670998.

In the ABI specification there should be a separate section explaining how solidity types are mapped to ABI types, instead of mentioning it "inline".
This section should include the fact that ``address payable`` is mapped to ``address`` in the ABI.

### PR DESCRIPTION
**Pull request by leonardoalt**
Closes #4950

### CODE REVIEW

File: docs/abi-spec.rst [hunk 1]
```diff
@@ -78,6 +78,18 @@ Types can be combined to a tuple by enclosing them inside parentheses, separated
...
+   - :ref:`enum<enums>`: maps to the smallest ``uint`` type that is large enough
+     to hold all values. For example, an ``enum`` of 255 values or less is
+     mapped to ``uint8`` and an ``enum`` of 256 values is mapped to ``uint16``.
+
+   - :ref:`contract<contracts>`: maps to ``address``.
+
+   - :ref:`address payable<address>`: maps to ``address``.
+
 .. note::
     Solidity supports all the types presented above with the same names with the exception of tuples. The ABI tuple type is utilised for encoding Solidity ``structs``.
```
Comment (axic) [inline docs/abi-spec.rst]: This line should be merged with the above.

However I may prefer a section at the end of the document listing a table of mappings?

File: docs/abi-spec.rst [hunk 2]
```diff
@@ -78,6 +78,18 @@ Types can be combined to a tuple by enclosing them inside parentheses, separated
 
 It is possible to form tuples of tuples, arrays of tuples and so on.  It is also possible to form zero-tuples (where ``n == 0``).
 
+.. note::
+   The following Solidity types are not part of the ABI, and are mapped to
+   some existing ABI type:
```
Comment (erak) [inline docs/abi-spec.rst]: Maybe replace ``some`` by ``an``?

**Comment by leonardoalt** [comment]
Changed it to table

File: docs/abi-spec.rst [hunk 3]
```diff
@@ -81,6 +81,23 @@ It is possible to form tuples of tuples, arrays of tuples and so on.  It is also
...
 
+Mapping Solidity to ABI types
+-----------------------------
+The following Solidity types (left column) are not part of the ABI, and are
+mapped to an existing ABI type (right column):
+
++-------------------------------+-----------------------------------------------------------------------------+
+|      Solidity                 |                                           ABI                               |
++===============================+=============================================================================+
+|:ref:`address payable<address>`|Maps to ``address``.                                                         |
```
Comment (chriseth) [inline docs/abi-spec.rst]: a construct that does not have a subject is not a sentence and thus should not start with uppercase and should not end with a period ;)

Perhaps the "maps to" could also be in the header?

**Comment by leonardoalt** [comment]
Fixed & rebased & squashed

**Review by chriseth — APPROVED** [review]

File: docs/abi-spec.rst [hunk 4]
```diff
@@ -81,6 +81,24 @@ It is possible to form tuples of tuples, arrays of tuples and so on.  It is also
 .. note::
     Solidity supports all the types presented above with the same names with the exception of tuples. The ABI tuple type is utilised for encoding Solidity ``structs``.
```
Comment (axic) [inline docs/abi-spec.rst:82]: Should have squashed this note into the table.

Comment (leonardoalt) [reply docs/abi-spec.rst:82]: Created #5024

**Pull request merged**